File: Crio/crio_programming_interview_problems-master/GenerateParentheses/GenerateParentheses.java
import java.util.Scanner;
import java.util.*;

class GenerateParentheses {
    // CRIO_SOLUTION_START_MODULE_L2_PROBLEMS

    void paren(String str, int left, int right, List<String> result) {

        if (left > right)
            return;

        if ((left == 0) && (right == 0)) {
            //System.out.println(str);
            result.add(str);
            return;
        }

        if (left != 0) {
            paren(str + "(", left-1, right, result);
        }
        if (right != 0) {
            paren(str + ")", left, right-1, result);
        }
    }
    // CRIO_SOLUTION_END_MODULE_L2_PROBLEMS

    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();

        // CRIO_SOLUTION_START_MODULE_L2_PROBLEMS
        paren("", n, n, result);
        // CRIO_SOLUTION_END_MODULE_L2_PROBLEMS

        return result;
    }

    public static void main(String args[]) {

        Scanner scanner = new Scanner(System.in);

        int n = scanner.nextInt();
        List<String> result = new GenerateParentheses().generateParenthesis(n);

        for (int i = 0; i < result.size(); ++i)
            System.out.println(result.get(i));
    }
}

File: Crio/crio_programming_interview_problems-master/GenerateParentheses/GenerateParentheses.cpp
#include <bits/stdc++.h>
using namespace std;

// CRIO_SOLUTION_START_MODULE_L2_PROBLEMS
void paren(vector<string> &ans,int left,int right,string curr){
    if(left>right){
        return;
    }
    if(left==0&&right==0){
        ans.push_back(curr);
        return;
    }

    if(left>0){
        string new_curr = curr;
        new_curr.push_back('(');
        paren(ans,left-1,right,new_curr);
    }
    if(right>0){
        string new_curr = curr;
        new_curr.push_back(')');
        paren(ans,left,right-1,new_curr);
    }
}
// CRIO_SOLUTION_END_MODULE_L2_PROBLEMS


vector<string> generateParentheses(int n){    
    // CRIO_SOLUTION_START_MODULE_L2_PROBLEMS
    vector<string> ans;
    paren(ans,n,n,"");
    return ans;
    // CRIO_SOLUTION_END_MODULE_L2_PROBLEMS

}

int main(){
    int n;cin>>n;

    vector<string> ans = generateParentheses(n);
    for(auto parentheses: ans ){
        cout<<parentheses<<endl;
    }
}